### PR TITLE
Share requests.Session across SFType instances

### LINKED ISSUE #91: SFType doesn't inherit session.
This effectively disables connection pooling and greatly hurts performance.

### PR DESCRIPTION
**Pull request by Trii**
Fixes #91 

I was a bit more explicit in the creation of `SFType` and specified parameter names instead of relying on their position. I hope that is OK.

### CODE REVIEW

**Comment by demianbrecht** [comment]
Hey @Trii, cheers for the patch!

Can you perhaps add a small unit test or two confirming the expected functionality? Offhand what I'm looking for specifically is to ensure that an instance of `SFType` is reusing the `requests.Session` of the `Salesforce` instance that's instantiating it.

Otherwise, LGTM. Unfortunately, I'm still blocked on merging PRs, but hoping to have that cleared up ASAP.

Thanks again!

**Comment by Trii** [comment]
Will do!

**Comment by Trii** [comment]
@demianbrecht tests for sharing a session and making sure proxies propagate through the `__getattr__`  calls are set. I also noticed that when `session=None`, `SalesforceLogin` was using `requests.post` but then afterwards, `Salesforce` was creating a new `requests.Session` and using it from then on out. I moved the logic up so that a new `requests.Session` is being passed to `SalesforceLogin` for consistency. [This change here](https://github.com/heroku/simple-salesforce/pull/98/files#diff-386c8e7f4cc62563b2b586356d1c67ffR76)

File: simple_salesforce/api.py [hunk 1]
```diff
@@ -442,11 +443,15 @@ def __init__(
...
         * proxies -- the optional map of scheme to proxy server
+        * session -- Custom requests session, created in calling code. This
+                     enables the use of requests Session features not otherwise
+                     exposed by simple_salesforce.
         """
         self.session_id = session_id
         self.name = object_name
-        self.request = requests.Session()
-        self.request.proxies = proxies
+        self.request = session or requests.Session()
```
Comment (demianbrecht) [inline simple_salesforce/api.py]: I see the inconsistent naming between this and the `Salesforce` object (`request` here but `session` in `Salesforce`). I would much rather see `session` used consistently between the two as it's what it's actually a reference to an instance of. Also, parameters and attributes will also be consistent.

File: simple_salesforce/api.py [hunk 2]
```diff
@@ -442,11 +443,15 @@ def __init__(
...
+        if proxies is not None:  # don't wipe out original proxies with None
+            self.request.proxies = proxies
```
Comment (demianbrecht) [inline simple_salesforce/api.py]: I think that if a session is passed in explicitly, `proxies` should be ignored. The side effect on the session instance owned by an upstream object concerns me a little.

**Comment by demianbrecht** [comment]
Hey @Trii, we're now unblocked on accepting contributions! I've made a couple comments on the PR, hopefully you're still interested in contributing after such a long time.

File: simple_salesforce/api.py (hunk 1 above)
Comment (demianbrecht) [inline simple_salesforce/api.py]: Note to self: Stop being off your face.

**Comment by Trii** [comment]
Hey @demianbrecht, I don't work with Salesforce any longer but if you'd like to merge it still I can address the comments and fix up the PR

**Comment by demianbrecht** [comment]
I was going to pick this up myself, but if you have the cycles to finish 'er up, that'd be awesome. Cheers!

**Comment by Trii** [comment]
Should be all set @demianbrecht. I ended up rebasing off of the current master since this PR was so old.

**Comment by demianbrecht** [comment]
Thanks for the quick turnaround @Trii! Made a couple slight modifications but great stuff.

**Comment by Trii** [comment]
You're welcome. Hope it helps!

**Pull request merged**